### PR TITLE
Build CI/CD Pipeline for GX Credential Generator

### LINKED ISSUE #76: Build CI/CD Pipeline for GX Credential Generator
We need a CI/CD Pipeline to increase/ensure code quality.  

Value-added tools:

- [ ] ~[black](https://github.com/psf/black)~ See https://github.com/SovereignCloudStack/gx-credential-generator/issues/76#issuecomment-2044811044
- [x]  [isort](https://pycqa.github.io/isort/)
- [x]  [flake8](https://flake8.pycqa.org/en/latest/)
- [x]  [code covarge](https://coverage.readthedocs.io/en/7.4.4/)
- [x]  Markdown Lint from [standard repo](https://github.com/SovereignCloudStack/standards/blob/main/.github/workflows/pr-markdownlint.yml)
- [x]  Markdown link validator from [standards repo](https://github.com/SovereignCloudStack/standards/blob/main/.github/workflows/link-validator.yml)

### PR DESCRIPTION
**Pull request by anjastrunk**
close #76

### CODE REVIEW

File: k8s-discovery.py [hunk 1]
```diff
@@ -183,7 +232,7 @@ def main(argv):
     if ofile == "/dev/stdout":
         print(my_kube, file=sys.stdout)
     else:
-        print(my_kube, file=open(ofile, 'a', encoding="UTF-8"))
+        print(my_kube, file=open(ofile, "a", encoding="UTF-8"))
```
Comment (martinmo) [inline k8s-discovery.py:235]:
```suggestion
        print(my_kube, file=open(ofile, mode="a", encoding="UTF-8"))
```

File: .github/workflows/build-python-app.yml [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+        run: |
+          python -m pip install --upgrade pip setuptools
+          pip install -r test-requirements.txt
+          pip install -r requirements.txt
+
+      - name: Test with pytest
+        run: |
+          coverage run --omit=tests/*,generator/common/gx_schema.py,*__init__*,generator/vendor/* -m pytest tests
+          coverage xml
+          coverage report -m
```
Comment (martinmo) [inline .github/workflows/build-python-app.yml:40]: When pytest is used, I think it is better to use the `pytest-cov` plugin for code coverage (it also uses `coverage` under the hood). See https://github.com/SovereignCloudStack/standards/tree/main/Tests#unit-and-regression-tests (Option C) for example.

I also recommend to put the configuration (flags such as `--omit`) into a `.coveragerc` file, so you have easy access to the same configuration if you test and measure coverage locally.

File: gx-cred-generator.py [hunk 3]
```diff
@@ -243,10 +278,10 @@ def main(argv):
         print(output(mycloud, myk8s), file=sys.stdout)
     else:
         if ostack.outjson:
-            file_name = f'{ofile}_{int(time())}.json'
+            file_name = f"{ofile}_{int(time())}.json"
         else:
-            file_name = f'{ofile}_{int(time())}.yaml'
-        print(output(mycloud, myk8s), file=open(file_name, 'a', encoding="UTF-8"))
+            file_name = f"{ofile}_{int(time())}.yaml"
+        print(output(mycloud, myk8s), file=open(file_name, "a", encoding="UTF-8"))
```
Comment (martinmo) [inline gx-cred-generator.py:284]:
```suggestion
        print(output(mycloud, myk8s), file=open(file_name, mode="a", encoding="UTF-8"))
```

File: tests/test_cli.py [hunk 4]
```diff
@@ -85,13 +87,33 @@ def test_openstack_empty(self, os_connect):
 
         self.assertIsNone(result.exception)
         self.assertEqual(0, result.exit_code)
-        with open(get_absolute_path("tests/data/empty_credential.json"), "r") as json_file:
+        with open(
+                get_absolute_path("tests/data/empty_credential.json"), "r"
```
Comment (martinmo) [inline tests/test_cli.py:91]:
```suggestion
                get_absolute_path("tests/data/empty_credential.json"), mode="r"
```

**Review by martinmo — CHANGES_REQUESTED** [review]
Looks good! I have found a few places which could be improved (maybe), please have a look at my comments.

FYI, the link-validator and build-python-app workflows have warnings wrt to deprecated actions (checkout and setup-python).

**Comment by anjastrunk** [comment]
@martinmo Thanks for your review. I replaced `coverage` with `pytest-cov` and bumped up github action to `actions/checkout@v4`

File: tests/coverage.xml
Comment (martinmo) [inline tests/coverage.xml:1]: This file shouldn't be in version control and should be ignored.

**Review by martinmo — CHANGES_REQUESTED** [review]

File: .coveragerc [hunk 5]
```diff
@@ -0,0 +1,3 @@
+[run]
+omit = *_test.py,tests/*,generator/common/gx_schema.py,*__init__*,generator/vendor/*
```
Comment (martinmo) [inline .coveragerc:2]:
```suggestion
omit = 
  tests/*
  generator/common/gx_schema.py
  generator/vendor/*
```

`coverage` supports this multi line format, which is easier to read. Also, some of the entries are unnecessary:

* there are no `*_test.py` files in the repo and
* there is only a single `__init__.py` file in the already omitted `generator/vendor/`. (If you really intend to exclude all future `__init__.py` files, I suggest to use the more specific `*/__init__.py` instead.)

File: .github/workflows/lint-python-syntax.yml [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+      - .github/workflows/lint-python-syntax.yml
+
+jobs:
+  lint-python-syntax:
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/checkout@v4
+      - uses: actions/setup-python@v4
+        with:
+          python-version: '3.x'
```
Comment (martinmo) [inline .github/workflows/lint-python-syntax.yml:16]:
```suggestion
          python-version: '3.10'
```

File: .github/workflows/lint-python-syntax.yml [hunk 7]
```diff
@@ -0,0 +1,29 @@
...
+      - run: pip3 install flake8
+      - run: flake8 --exclude generator/common/gx_schema.py
+
+  check-sorted-imports:
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/checkout@v3
+      - uses: actions/setup-python@v4
+        with:
+          python-version: '3.x'
```
Comment (martinmo) [inline .github/workflows/lint-python-syntax.yml:26]:
```suggestion
          python-version: '3.10'
```

File: flake8 [hunk 8]
```diff
@@ -0,0 +1,3 @@
+[flake8]
+ignore = E501, E251, E226, E221
+
```
Comment (martinmo) [inline flake8:3]: This file is a duplicate of the hidden `.flake8` file with the same contents and can be removed.

**Review by martinmo — CHANGES_REQUESTED** [review]
In the last review I forgot to ask: is the append mode in some of the `open(…)` calls intentional, instead of `mode="w"`? I know it is not actually the scope of this PR, but because of the reformatting I noticed these in the diffs.

It is not part of the diff, but can you change the hidden `.flake8` to just ignore `E501` (line length) as follows:

```
[flake8]
ignore = E501
```

The other warnings should remain active and the code conforms to them, at the moment.

Apart from that I only found some minor nitpicks.

File: flake8 (hunk 8 above)
Comment (anjastrunk) [reply flake8]: Removed. See 97a3f73dfe9cf12024038ce16547b82819d454ca

**Comment by anjastrunk** [comment]
> In the last review I forgot to ask: is the append mode in some of the `open(…)` calls intentional, instead of `mode="w"`? I know it is not actually the scope of this PR, but because of the reformatting I noticed these in the diffs.

I have no idea, if append mode is intended. As I didn't write this code, I will leave it as it is. BTW: Old code will be removed/refactored in later PRs.

File: .flake8 [hunk 9]
```diff
@@ -1,2 +1,2 @@
 [flake8]
-ignore = E501,E226,E251,E221
+ignore = E501,E226,E251,E221,E501
```
Comment (martinmo) [inline .flake8:2]:
```suggestion
ignore = E501
```

is enough and the lint still passes with these settings (see https://github.com/SovereignCloudStack/gx-credential-generator/pull/84#pullrequestreview-2011643696)

**Review by martinmo — CHANGES_REQUESTED** [review]
Just one suggestion for the `.flake8` file, otherwise okay! :+1:

**Review by martinmo — APPROVED** [review]

**Pull request merged**